### PR TITLE
Support destructuring private proposal

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Just enables `destructuringPrivate` plugin and adds tests.

Partial of #12255

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
Can you mention that we support this https://github.com/tc39/proposal-regexp-set-notation/ in changelog? I forgot to add changelog in #12241, unless you want add seperate changelog for it.

**Comment by fisker** [comment]
How this line affect? https://github.com/prettier/prettier/blob/34fbd099dee06a8b4d68684ccaf3cd5340c08583/src/language-js/print/assignment.js#L378

File: tests/format/js/babel-plugins/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -626,6 +626,75 @@ function enumerable(value) {
...
+=====================================output=====================================
+class Foo {
+  #x = 1;
+  constructor() {
+    console.log(this.#x); // => 1
+    const { #x: x } = this;
+    console.log(x); // => 1
+  }
+  equals({ #x: otherX }) {
+    const { #x: currentX } = this;
```
Comment (fisker) [inline tests/format/js/babel-plugins/__snapshots__/jsfmt.spec.js.snap:690]: Is assign allowed in this syntax?

**Comment by sosukesuzuki** [comment]
> Can you mention that we support this https://github.com/tc39/proposal-regexp-set-notation/ in changelog? I forgot to add changelog in https://github.com/prettier/prettier/pull/12241, unless you want add seperate changelog for it.

I prefer to separate changelog so I'll create new PR that adds changelog for the proposal unicode set notation.

File: tests/format/js/babel-plugins/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (sosukesuzuki) [reply tests/format/js/babel-plugins/__snapshots__/jsfmt.spec.js.snap:690]: Yes. This example is copied from the example of https://github.com/tc39/proposal-destructuring-private.

**Comment by sosukesuzuki** [comment]
> How this line affect?

This line only affects the RHS (only nodes in inner of `CallExpression`) of the assignment, so I don't think it is relevant to the syntax of Destructuring Private. (`{ #foo: foo }` can appear on only LHS.)

File: tests/format/js/babel-plugins/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (fisker) [reply tests/format/js/babel-plugins/__snapshots__/jsfmt.spec.js.snap:690]:
```js
({ #x: variable.or.property.to.assign } = this);
```

Comment (sosukesuzuki) [reply tests/format/js/babel-plugins/__snapshots__/jsfmt.spec.js.snap:690]: Ah sorry I misunderstood. Destructuring private in assign expression is maybe invalid.

Comment (sosukesuzuki) [reply tests/format/js/babel-plugins/__snapshots__/jsfmt.spec.js.snap:690]: @fisker Maybe this is valid. So what should we do for it?

Comment (fisker) [reply tests/format/js/babel-plugins/__snapshots__/jsfmt.spec.js.snap:690]: Seems working fine, 

**Prettier pr-12276**
[Playground link](https://deploy-preview-12276--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAucAbAhgZ0wAgII7AA6UOOJZAtnDABYQAmAFAJSEVk5PA4DEAHkhwA3dACcAlugBGqOADoIY+QAcxEFXDEwAnvJgR5WTBIDmpAL44AvDjoTMLANwcyFkhZAAaEBpgToTGRQcXUAdwAFcQQglHRUMPQdIJ9pMXQwAGsaAGV0agAZCSg4ZAAzeMw4VPSs3JUM4tNkGDEAV2qQOEppOAYGPoL0KFM29FM4ADElSnQYfxHkEHQ2g28QWhhKVAB1Wgl4TAawOByYg4lhA50lsGx14qrtCPTTWfLKzoArTH4cprkAEU2hB4B9UFUfA0xE8ltIZHBUOs1MUYDsJAw6MgABwABih6iqO3SKiWajgT2EpR8AEcQfAXhpYstMABaEp9PrrMRwOkSHkvcbvJAVCGdKqUCQtdrigFwYGg0oiz4+GAydGY2jIABMqvSElQTQAwhBKMKupgAKzrNpVAAqMlioshIGEHQAklABrAcmBJCoYHgvTldHJwVULBYgA)

```sh
--parser babel
```

**Input:**

```jsx
class A {
  
  method() {
    ({ #x: variable.or.property.to.assign } = this);
    }
}
```

**Output:**

```jsx
class A {
  method() {
    ({ #x: variable.or.property.to.assign } = this);
  }
}

```

Add a test for it, and check if we need adjust the printer.

This syntax is a good one, I was wait for it for quite a while, I may actual use it.

Comment (fisker) [reply tests/format/js/babel-plugins/__snapshots__/jsfmt.spec.js.snap:690]: Yes, valid. https://github.com/tc39/proposal-destructuring-private/issues/4#issue-1045243014

Comment (fisker) [reply tests/format/js/babel-plugins/__snapshots__/jsfmt.spec.js.snap:690]: Don't forget add some tests for destructuring with default value.

**Review by fisker — APPROVED** [review]

**Pull request merged**